include/envoy/stats/timespan.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/time.h"
+#include "envoy/stats/stats.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * An individual timespan that flushes its measured value to a Duration histogram. The initial time
+ * is captured on construction. A timespan must be completed via complete() for it to be stored. If
+ * the timespan is deleted this will be treated as a cancellation.
+ */
+class Timespan {
+public:
+  Timespan(Histogram& histogram) : histogram_(histogram), start_(std::chrono::steady_clock::now()) {
+    if (histogram.type() != Histogram::ValueType::Duration) {
+      throw EnvoyException("Cannot intialize a timespan with a non-time valued histogram");
+    }
+  }
+
+  virtual ~Timespan() {}
+
+  /**
+   * Complete the timespan and send the time to the histogram.
+   */
+  virtual void complete() {
+    histogram_.recordValue(std::chrono::duration_cast<std::chrono::milliseconds>(
```

Comment:
Should this stuff be in the `source/common` implementation rather than interface?

Reply:
As was discussed in #1751, this eliminates the problem of everything that wants to create a `Timespan` that flushes to a histogram having to depend on some stats implementation lib rather than just the set of stats interfaces (everything else in stats currently works this way outside of the creation of the store).  And the implementation if this class is relatively small.  I'm cool changing it if we're okay with those tradeoffs.